model-optimizer/extensions/ops/elementwise.py

```diff
@@ -22,13 +23,49 @@ def override_data_type_of_constant(node: Node):
         # power value 2, or when replacing Neg operation with Mul with -1 as second input.
         in_node_0 = node.in_port(0).get_source().node
         in_node_1 = node.in_port(1).get_source().node
-        if in_node_0.op == 'Const':
-            convert_const_node_value_type(in_node_0, in_type_1)
-        elif in_node_1.op == 'Const':
-            convert_const_node_value_type(in_node_1, in_type_0)
+
+        if in_node_0.soft_get('op') != 'Const' and in_node_1.op != 'Const':
+            raise Exception("Elementwise operation '{}' has inputs of different data types: '{}' and '{}' "
+                            'that cannot be aligned'.format(node.soft_get('name'), in_type_0, in_type_1))
+
+        if in_node_0.soft_get('op') == 'Const':
+            node_to_convert, src_type, dst_type = in_node_0, in_type_0, in_type_1
+        else:
+            node_to_convert, src_type, dst_type = in_node_1, in_type_1, in_type_0
+        log.error("Changing Const node '{}' data type from {} to {} for Elementwise operation".format(
+            node.soft_get('name', node_to_convert.id), src_type, dst_type),
+            extra={'is_warning': True})
+        convert_const_node_value_type(node_to_convert, dst_type)
+
+
+def override_input_types_of_pow(pow_node: Node):
+    # according to ONNX specification Pow op can have different input types
+    # https://github.com/onnx/onnx/blob/master/docs/Operators.md#Pow (Examples/types)
+
+    base_type = pow_node.in_port(0).get_data_type()
+    exponent_type = pow_node.in_port(1).get_data_type()
+    if type(base_type) != np.dtype:
+        base_type = np.dtype(base_type)
+    if type(exponent_type) != np.dtype:
+        exponent_type = np.dtype(exponent_type)
+
+    prefix = pow_node.soft_get('name', pow_node.id) + '/Cast_{}_to_{}_type'
+    if base_type != exponent_type:
+        if base_type.itemsize >= exponent_type.itemsize:
```

Comment:
This is wrong approach. Consider the following example:
```
>>> np.dtype(np.int64).itemsize >= np.dtype(np.float32).itemsize
True
```

The size of `int64` is bigger than the size of `fp32` but in fact the `int64` must be converted to `float` since the floating point value is wider than the `int64` value (I believe that the term "wider" means wider range of values and not the size of the data type). Please, double check this using the ONNX layer tests.

Reply:
when I was writing solution I was intended that it behaves in a such way that in the example above base will not be converted into `float32`.

Unfortunately layer tests can not help us in this situation, we need to look onnx runtime behavior. ONNX spec only says that output will of `int64` type. The approach with casting is just how onnx_importer relaxed input types constrains. I implemented exactly the same as them, and according to their code `wider` for them meant wider size.
https://github.com/openvinotoolkit/openvino/blob/master/ngraph/frontend/onnx_import/src/op/pow.cpp#L30-L46

Considering that we also have difficulties with IR reader check after this Pow type Casting, I propose to revert **problem-3** related changes since there left quite a little time to detailly check our implementation and I'm not quite sure that onnx_importer solution is correct as well.

Moreover, it's not a bug but rather feature support and ewe passed 2021.4 FF. It looked that we can easily add this but it turns out that everything is not so simple.

Reply:
reverted these changes for "problem-3*